datastore/http/http.go

```diff
@@ -0,0 +1,174 @@
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/teslamotors/fleet-telemetry/metrics"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter"
+	"github.com/teslamotors/fleet-telemetry/telemetry"
+)
+
+// Producer client to handle http
+type Producer struct {
+	namespace        string
+	metricsCollector metrics.MetricCollector
+	logger           *logrus.Logger
+	records          chan *telemetry.Record
+	address          string
+	httpClient       http.Client
+}
+
+// Metrics stores metrics reported from this package
+type Metrics struct {
+	produceCount adapter.Counter
+	byteTotal    adapter.Counter
+	errorCount   adapter.Counter
+	bufferSize   adapter.Gauge
+}
+
+// Config contains the data necessary to configure an http producer.
+type Config struct {
+	// WorkerCount is the number of http producer routines running.
+	// This number should be increased if `http_produce_buffer_size` is growing.
+	WorkerCount int `json:"worker_count"`
+
+	// Address is the address to produce requests to.
+	Address string `json:"address"`
+
+	// Timeout is the number of seconds to wait for a response. Defaults to 10.
+	Timeout int `json:"timeout"`
+}
+
+const (
+	statusCodePreSendErr = "NOT_SENT"
+	bufferSize           = 128
+)
+
+var (
+	metricsRegistry Metrics
+	metricsOnce     sync.Once
+)
+
+// NewProducer sets up an HTTP producer
+func NewProducer(config *Config, metricsCollector metrics.MetricCollector, namespace string, logger *logrus.Logger) (telemetry.Producer, error) {
+	registerMetricsOnce(metricsCollector)
+
+	err := validateConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	requests := make(chan *telemetry.Record, bufferSize)
+
+	producer := &Producer{
+		namespace:        namespace,
+		metricsCollector: metricsCollector,
+		records:          requests,
+		address:          config.Address,
+		logger:           logger,
+		httpClient: http.Client{
+			Timeout: time.Duration(config.Timeout) * time.Second,
+		},
+	}
+
+	for i := 0; i < config.WorkerCount; i++ {
+		go producer.worker(requests)
+	}
+
+	logger.Infof("registered http producer for namespace: %s", namespace)
+	return producer, nil
+}
+
+// validateConfig validates configuration values and sets defaults if value not set
+func validateConfig(c *Config) error {
+	if c.WorkerCount < 0 {
+		return errors.New("invalid http worker count")
+	}
+	if c.WorkerCount == 0 {
+		c.WorkerCount = 5
+	}
+	if c.Timeout < 0 {
+		return errors.New("invalid http timeout")
+	}
+	if c.WorkerCount == 0 {
+		c.Timeout = 10
+	}
+	return nil
+}
+
+func (p *Producer) worker(records <-chan *telemetry.Record) {
+	for record := range records {
+		metricsRegistry.bufferSize.Sub(1, map[string]string{})
+		p.sendHTTP(record)
+	}
+}
+
+func (p *Producer) sendHTTP(record *telemetry.Record) {
+	url := fmt.Sprintf("%s?namespace=%s&type=%s", p.address, p.namespace, record.TxType)
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(record.Payload()))
```

Comment:
we should also send record metadata https://github.com/teslamotors/fleet-telemetry/blob/main/telemetry/record.go#L63